### PR TITLE
Create article Authenticating with JWT token

### PR DESCRIPTION
**Pull request by janan07**
First draft of Using JWT token with API Mediation Layer for review

### CODE REVIEW

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+1. The API client obtains a JWT token by using the POST method on the `/auth/login` endpoint of the API service that requires a valid user ID and password.
+
+2. The API service then calls an *authentication provider* that returns a JWT token that contains the user ID claim in the HTTP cookie named `apimlAuthenticationToken` with attributes `HttpOnly` and `Secure`.
+
+3. The API client stores the JWT token or cookie and sends the token with every request as a cookie with the name `apimlAuthenticationToken`.
+
+#### Obtaining the token
+
+<!-- We should outline how to obtain a token in a sequence of numbered steps. -->
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:28]: There is one step - call the endpoint with the credentials. Either basic authentication or client certificate.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+- The full URL is the base URL of the API service plus `/auth/login`. If the application has the base URL with `/api/v1`, the full URL could have the format: `https://hostname:port/api/v1/auth/login`.
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:30]: The full path for API ML will be /gateway/auth/login - The SDK has a bit of difference.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    "password": "zowe"
+}
+
+HTTP/1.1 204
+Set-Cookie: apimlAuthenticationToken=eyJhbGciOiJSUzI1NiJ9...; Path=/; Secure; HttpOnly
+```
+
+#### Authenticated request
+
+One option <!-- One option to do what? --> is for the API client to pass the JWT token as a cookie header with the name `apimlAuthenticationToken`:
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:69]: One option to send a JWT token with the request

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+Content-Type: application/json;charset=UTF-8
+
+{
+    "userId": "zowe",
+    "creation": "2019-11-29T13:39:18.000+0000",
+    "expiration": "2019-11-30T13:39:18.000+0000"
+}
+```
+
+<!-- It seems that the following section is a work in progress and is not ready for the published documentation -->
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:137]: I agree, I just started with the transfer of the information about the client usage to this page.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+```json
+{
+  "sub": "zowe",
+  "iat": 1575034758,
+  "exp": 1575121158,
+  "iss": "Zowe Sample API Service",
+  "jti": "ac2eb63e-caa6-4ccf-a527-95cb61ad1646"
+}
+```
+### Support for token-based authentication in the Zowe REST API SDK
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:197]: This should be removed, it applies to extend section.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+When a client authenticates with the API ML, the client receives the JWT token whch can then be used for further authentication. If z/OSMF is configured as the authentication provider and the client already received a JWT token produced by z/OSMF, it is possible to reuse this token within API ML for authentication.
+
+This parent article describes how services in the Zowe API ecosystem are expected to accept and use JWT tokens so that API clients have a stadardized experience. 
+
+By default, JWT tokens are produced by z/OSMF and the API Mediation Layer serves only as a proxy. For information about how to change who and how tokens are produced, see [Authentication Providers within Enable Single Sign On for Clients](../user-guide/api-mediation/configuration-jwt/#saf-as-an-authentication-provider)
+
+<!-- moved https://docs.zowe.org/stable/extend/extend-apiml/authentication-for-apiml-services/#authentication-providers to the https://docs.zowe.org/stable/user-guide/api-mediation/configuration-jwt
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:16]: Thanks.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+<!-- moved https://docs.zowe.org/stable/extend/extend-apiml/authentication-for-apiml-services/#authentication-providers to the https://docs.zowe.org/stable/user-guide/api-mediation/configuration-jwt 
+-->
+
+### Token-based Login Flow and Request/Response Format
+
+The following sequence describes how authentication through JWT tokens works:
+
+1. The API client obtains a JWT token by using the POST method on the `/auth/login` endpoint of the API service that requires a valid user ID and password.
+
+2. The API service then calls an *authentication provider* that returns a JWT token that contains the user ID claim in the HTTP cookie named `apimlAuthenticationToken` with attributes `HttpOnly` and `Secure`.
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:25]: JUst remove the point 2. 

Point 1 is followed by 3.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+3. The API client stores the JWT token or cookie and sends the token with every request as a cookie with the name `apimlAuthenticationToken`.
+
+## Obtaining a token
+
+To obtain a token, call the endpoint with the credentials for either basic authentication or the client certificate.
+
+<!-- Check the previous statement for accuracy. -->
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:33]: Yes, that's correct other than for my comment on previous section.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+:::tip
+The first option (using a cookie header) is recommended for web browsers with the attributes `Secure` and `HttpOnly`.
+Browsers store and send cookies automatically.
+Cookies are present on all requests, including those coming from DOM elements, and are compatible with web mechanisms such as CORS, SSE, or WebSockets.
+
+Cookies are more diffcult to support in non-web applications.
+Headers, such as `Authorization: Bearer`, can be used in non-web applications. Such headers, however, are difficult to use and secure in a web browser.
+The web application needs to store these headers and attach these headers to all requests where headers are required.
+
+Note that the API service needs to support both of them <!-- What is "both of them" --> so the API clients can use what makes more sense for them.
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:89]: Remov this line. It's not relevant for this article. It relates to the extensions.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+These fields correspond to `iss`, `exp`, and `sub` JWT token claims. The timestamps are in ISO 8601 format.
+
+<!-- We should describe what the following command does, such as "Execute the following curl command to check the status of your token." -->
+**Example:**
+
+```bash
+curl -k --cookie "apimlAuthenticationToken=eyJhbGciOiJSUzI1NiJ9..." -X GET "https://localhost:10080/api/v1/auth/query"
+```
+
+The following output describes the status of the token: <!-- Please validate this statement. -->
```
Comment (balhar-jakub) [inline docs/user-guide/using-jwt-token-with-apiml.md:138]: That is correct. Status of the JWT token.

File: docs/user-guide/using-jwt-token-with-apiml.md [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+- The refresh request requires the token in one of the following formats:
+  - Cookie named `apimlAuthenticationToken`.
+  - Bearer authentication
+:::
+
+For more information, see the OpenAPI documentation of the API Mediation Layer in the API Catalog.
+
+The following request receives a valid JWT token and returns the new valid JWT token. As such, the expiration time is reset. 
+
+```bash
```
Comment (taban03) [inline docs/user-guide/using-jwt-token-with-apiml.md:172]:
```bash
curl -v -c - -X POST --data '{"username":"zowe","password":"zowe"}' "https://localhost:10080/api/v1/auth/refresh"
```

**Review by taban03 — APPROVED** [review]

**Review by balhar-jakub — APPROVED** [review]

**Pull request merged**